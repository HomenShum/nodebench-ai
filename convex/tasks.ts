import { v } from "convex/values";
import { mutation, query } from "./_generated/server";
import type { Id } from "./_generated/dataModel";
import { getAuthUserId } from "@convex-dev/auth/server";

// Helper to normalize/validate user id, consistent with folders.ts pattern
async function getSafeUserId(ctx: any): Promise<Id<"users">> {
  const rawUserId = await getAuthUserId(ctx);
  if (!rawUserId) throw new Error("Not authenticated");
  let userId: Id<"users">;
  if (typeof rawUserId === "string" && rawUserId.includes("|")) {
    const first = rawUserId.split("|")[0];
    if (!first || first.length < 10) throw new Error("Invalid user ID format. Please sign out and in.");
    userId = first as Id<"users">;
  } else {
    userId = rawUserId as Id<"users">;
  }
  const user = await ctx.db.get(userId);
  if (!user) throw new Error("User not found. Please sign out and sign back in.");
  return userId;
}

// Helper that does NOT throw for unauthenticated access; returns null instead.
async function getOptionalUserId(ctx: any): Promise<Id<"users"> | null> {
  const rawUserId = await getAuthUserId(ctx);
  if (!rawUserId) return null;
  let userId: Id<"users"> | null = null;
  if (typeof rawUserId === "string" && rawUserId.includes("|")) {
    const first = rawUserId.split("|")[0];
    if (!first || first.length < 10) return null;
    userId = first as Id<"users">;
  } else {
    userId = rawUserId as Id<"users">;
  }
  const user = userId ? await ctx.db.get(userId) : null;
  if (!user) return null;
  return userId;
}

// Create a new task
export const createTask = mutation({
  args: {
    title: v.string(),
    description: v.optional(v.string()),
    descriptionJson: v.optional(v.string()),
    status: v.optional(v.union(
      v.literal("todo"),
      v.literal("in_progress"),
      v.literal("done"),
      v.literal("blocked"),
    )),
    priority: v.optional(v.union(
      v.literal("low"),
      v.literal("medium"),
      v.literal("high"),
      v.literal("urgent"),
    )),
    dueDate: v.optional(v.number()),
    startDate: v.optional(v.number()),
    documentId: v.optional(v.id("documents")),
    eventId: v.optional(v.id("events")),
    assigneeId: v.optional(v.id("users")),
    refs: v.optional(
      v.array(
        v.union(
          v.object({ kind: v.literal("document"), id: v.id("documents") }),
          v.object({ kind: v.literal("task"), id: v.id("tasks") }),
          v.object({ kind: v.literal("event"), id: v.id("events") }),
        ),
      ),
    ),
    tags: v.optional(v.array(v.string())),
    color: v.optional(v.string()),
    order: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    const userId = await getSafeUserId(ctx);
    const now = Date.now();
    const taskId = await ctx.db.insert("tasks", {
      userId,
      title: args.title,
      description: args.description,
      descriptionJson: args.descriptionJson,
      status: args.status ?? "todo",
      priority: args.priority,
      dueDate: args.dueDate,
      startDate: args.startDate,
      documentId: args.documentId,
      eventId: args.eventId,
      assigneeId: args.assigneeId,
      refs: args.refs,
      tags: args.tags,
      color: args.color,
      order: args.order,
      createdAt: now,
      updatedAt: now,
    });
    return taskId;
  },
});

// Tiny resolver: get task title by id (authorized)
export const getTitle = query({
  args: { taskId: v.id("tasks") },
  handler: async (ctx, args) => {
    const userId = await getOptionalUserId(ctx);
    if (!userId) return null;
    const t = await ctx.db.get(args.taskId);
    if (!t || t.userId !== userId) return null;
    return { title: t.title } as { title: string };
  },
});

// Batch resolver: get task titles by ids (authorized)
export const getTitles = query({
  args: { ids: v.array(v.id("tasks")) },
  handler: async (ctx, args) => {
    const userId = await getOptionalUserId(ctx);
    if (!userId) return [] as Array<{ _id: Id<"tasks">; title: string }>;
    const out: Array<{ _id: Id<"tasks">; title: string }> = [];
    for (const id of args.ids) {
      const t = await ctx.db.get(id);
      if (t && t.userId === userId) {
        out.push({ _id: id, title: t.title });
      }
    }
    return out;
  },
});

// List tasks for current user filtered by assigneeId
export const listTasksByAssignee = query({
  args: { assigneeId: v.id("users") },
  handler: async (ctx, args) => {
    const userId = await getOptionalUserId(ctx);
    if (!userId) return [];
    return await ctx.db
      .query("tasks")
      .withIndex("by_user_assignee", (q: any) => q.eq("userId", userId).eq("assigneeId", args.assigneeId))
      .order("asc")
      .collect();
  },
});

// Toggle favorite on a task (owner-only)
export const toggleFavorite = mutation({
  args: { taskId: v.id("tasks") },
  returns: v.object({ isFavorite: v.boolean() }),
  handler: async (ctx, args) => {
    const userId = await getSafeUserId(ctx);
    const existing = await ctx.db.get(args.taskId);
    if (!existing) throw new Error("Task not found");
    if (existing.userId !== userId) throw new Error("Not authorized");

    const next = !existing.isFavorite;
    await ctx.db.patch(args.taskId, { isFavorite: next, updatedAt: Date.now() });
    return { isFavorite: next };
  },
});

// List recent tasks ordered by updatedAt descending
export const listTasksByUpdatedDesc = query({
  args: { limit: v.optional(v.number()) },
  handler: async (ctx, args) => {
    const userId = await getOptionalUserId(ctx);
    if (!userId) return [];
    const limit = Math.min(Math.max(1, args.limit ?? 10), 100);
    const rows = await ctx.db
      .query("tasks")
      .withIndex("by_user_updatedAt", (q: any) => q.eq("userId", userId))
      .order("desc")
      .take(limit);
    return rows;
  },
});

// Update a task (owner-only)
export const updateTask = mutation({
  args: {
    taskId: v.id("tasks"),
    title: v.optional(v.string()),
    description: v.optional(v.string()),
    descriptionJson: v.optional(v.string()),
    status: v.optional(v.union(
      v.literal("todo"),
      v.literal("in_progress"),
      v.literal("done"),
      v.literal("blocked"),
    )),
    priority: v.optional(v.union(
      v.literal("low"),
      v.literal("medium"),
      v.literal("high"),
      v.literal("urgent"),
    )),
    dueDate: v.optional(v.number()),
    startDate: v.optional(v.number()),
    documentId: v.optional(v.id("documents")),
    eventId: v.optional(v.id("events")),
    assigneeId: v.optional(v.id("users")),
    refs: v.optional(
      v.array(
        v.union(
          v.object({ kind: v.literal("document"), id: v.id("documents") }),
          v.object({ kind: v.literal("task"), id: v.id("tasks") }),
          v.object({ kind: v.literal("event"), id: v.id("events") }),
        ),
      ),
    ),
    tags: v.optional(v.array(v.string())),
    color: v.optional(v.string()),
    order: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    const userId = await getSafeUserId(ctx);
    const existing = await ctx.db.get(args.taskId);
    if (!existing) throw new Error("Task not found");
    if (existing.userId !== userId) throw new Error("Not authorized");

    const updates: any = { updatedAt: Date.now() };
    if (args.title !== undefined) updates.title = args.title;
    if (args.description !== undefined) updates.description = args.description;
    if (args.descriptionJson !== undefined) updates.descriptionJson = args.descriptionJson;
    if (args.status !== undefined) updates.status = args.status;
    if (args.priority !== undefined) updates.priority = args.priority;
    if (args.dueDate !== undefined) updates.dueDate = args.dueDate;
    if (args.startDate !== undefined) updates.startDate = args.startDate;
    if (args.documentId !== undefined) updates.documentId = args.documentId;
    if (args.eventId !== undefined) updates.eventId = args.eventId;
    if (args.assigneeId !== undefined) updates.assigneeId = args.assigneeId;
    if (args.refs !== undefined) updates.refs = args.refs;
    if (args.tags !== undefined) updates.tags = args.tags;
    if (args.color !== undefined) updates.color = args.color;
    if (args.order !== undefined) updates.order = args.order;

    await ctx.db.patch(args.taskId, updates);
    return { success: true };
  },
});

// Delete a task (owner-only)
export const deleteTask = mutation({
  args: { taskId: v.id("tasks") },
  handler: async (ctx, args) => {
    const userId = await getSafeUserId(ctx);
    const existing = await ctx.db.get(args.taskId);
    if (!existing) return { success: true };
    if (existing.userId !== userId) throw new Error("Not authorized");
    await ctx.db.delete(args.taskId);
    return { success: true };
  },
});

// Get a task
export const getTask = query({
  args: { taskId: v.id("tasks") },
  handler: async (ctx, args) => {
    const userId = await getOptionalUserId(ctx);
    if (!userId) return null;
    const task = await ctx.db.get(args.taskId);
    if (!task || task.userId !== userId) return null;
    return task;
  },
});

// List tasks by status
export const listTasksByStatus = query({
  args: { status: v.optional(v.union(
    v.literal("todo"),
    v.literal("in_progress"),
    v.literal("done"),
    v.literal("blocked"),
  )) },
  handler: async (ctx, args) => {
    const userId = await getOptionalUserId(ctx);
    if (!userId) return [];
    if (!args.status) return [];
    return await ctx.db
      .query("tasks")
      .withIndex("by_user_status", (q: any) => q.eq("userId", userId).eq("status", args.status))
      .order("asc")
      .collect();
  },
});

// List tasks due in [start, end]
export const listTasksDueInRange = query({
  args: { start: v.number(), end: v.number() },
  handler: async (ctx, args) => {
    const userId = await getOptionalUserId(ctx);
    if (!userId) return [];
    return await ctx.db
      .query("tasks")
      .withIndex("by_user_dueDate", (q: any) =>
        q.eq("userId", userId).gte("dueDate", args.start).lte("dueDate", args.end)
      )
      .order("asc")
      .collect();
  },
});

// List tasks by priority
export const listTasksByPriority = query({
  args: { priority: v.union(
    v.literal("low"),
    v.literal("medium"),
    v.literal("high"),
    v.literal("urgent"),
  ) },
  handler: async (ctx, args) => {
    const userId = await getOptionalUserId(ctx);
    if (!userId) return [];
    return await ctx.db
      .query("tasks")
      .withIndex("by_user_priority", (q: any) => q.eq("userId", userId).eq("priority", args.priority))
      .order("asc")
      .collect();
  },
});

// Convenience: list tasks due today, using timezone offset minutes (default 0)
export const listTasksDueToday = query({
  args: { tzOffsetMinutes: v.optional(v.number()) },
  handler: async (ctx, args) => {
    const userId = await getOptionalUserId(ctx);
    if (!userId) return [];
    const offsetMs = (args.tzOffsetMinutes ?? 0) * 60 * 1000;
    const now = Date.now();
    const localNow = now + offsetMs;
    const d = new Date(localNow);
    d.setUTCHours(0, 0, 0, 0);
    const start = d.getTime() - offsetMs;
    const end = start + 24 * 60 * 60 * 1000 - 1;

    return await ctx.db
      .query("tasks")
      .withIndex("by_user_dueDate", (q: any) => q.eq("userId", userId).gte("dueDate", start).lte("dueDate", end))
      .order("asc")
      .collect();
  },
});

// Convenience: list tasks due this week (Mon-Sun), using timezone offset minutes (default 0)
export const listTasksDueThisWeek = query({
  args: { tzOffsetMinutes: v.optional(v.number()) },
  handler: async (ctx, args) => {
    const userId = await getOptionalUserId(ctx);
    if (!userId) return [];
    const offsetMs = (args.tzOffsetMinutes ?? 0) * 60 * 1000;
    const now = Date.now();
    const localNow = now + offsetMs;
    const d = new Date(localNow);
    // Compute Monday as first day of week
    const day = d.getUTCDay(); // 0=Sun..6=Sat
    const diffToMonday = ((day + 6) % 7); // days since Monday
    const monday = new Date(localNow - diffToMonday * 24 * 60 * 60 * 1000);
    monday.setUTCHours(0, 0, 0, 0);
    const start = monday.getTime() - offsetMs;
    const end = start + 7 * 24 * 60 * 60 * 1000 - 1;

    return await ctx.db
      .query("tasks")
      .withIndex("by_user_dueDate", (q: any) => q.eq("userId", userId).gte("dueDate", start).lte("dueDate", end))
      .order("asc")
      .collect();
  },
});

// Move task to a new status and optionally update order (Kanban)
export const moveTask = mutation({
  args: {
    taskId: v.id("tasks"),
    status: v.union(
      v.literal("todo"),
      v.literal("in_progress"),
      v.literal("done"),
      v.literal("blocked"),
    ),
    order: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    const userId = await getSafeUserId(ctx);
    const existing = await ctx.db.get(args.taskId);
    if (!existing) throw new Error("Task not found");
    if (existing.userId !== userId) throw new Error("Not authorized");

    await ctx.db.patch(args.taskId, {
      status: args.status,
      order: args.order ?? existing.order,
      updatedAt: Date.now(),
    });
    return { success: true };
  },
});

// Rebalance orders in a lane (or all lanes) for the current user
export const rebalanceOrders = mutation({
  args: {
    status: v.optional(
      v.union(
        v.literal("todo"),
        v.literal("in_progress"),
        v.literal("done"),
        v.literal("blocked"),
      ),
    ),
  },
  returns: v.number(),
  handler: async (ctx, args) => {
    const userId = await getSafeUserId(ctx);
    const lanes: Array<"todo" | "in_progress" | "done" | "blocked"> = args.status
      ? [args.status]
      : ["todo", "in_progress", "done", "blocked"];

    let updated = 0;
    const STEP = 1000;

    for (const lane of lanes) {
      // Fetch all tasks for this user and status
      const rows = await ctx.db
        .query("tasks")
        .withIndex("by_user_status", (q: any) => q.eq("userId", userId).eq("status", lane))
        .order("asc")
        .collect();

      // Sort by existing order (missing orders go last), tie-breaker by createdAt
      rows.sort((a: any, b: any) => {
        const ao = typeof a?.order === "number" ? a.order : Number.MAX_SAFE_INTEGER;
        const bo = typeof b?.order === "number" ? b.order : Number.MAX_SAFE_INTEGER;
        if (ao !== bo) return ao - bo;
        const ac = typeof a?.createdAt === "number" ? a.createdAt : 0;
        const bc = typeof b?.createdAt === "number" ? b.createdAt : 0;
        return ac - bc;
      });

      // Assign spaced integer orders
      for (let i = 0; i < rows.length; i++) {
        const desired = (i + 1) * STEP;
        if (rows[i].order !== desired) {
          await ctx.db.patch(rows[i]._id, { order: desired, updatedAt: Date.now() });
          updated++;
        }
      }
    }

    return updated;
  },
});
