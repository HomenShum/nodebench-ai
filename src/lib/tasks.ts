import type { Pill } from "./metaPillMappers";

export type ShapedTask = {
  title?: string;
  dueAt?: string; // keep as ISO/string for current mapper
  project?: string;
  link?: string;
  updatedAt?: string | number | Date;
  priority?: 'low' | 'med' | 'high' | 'urgent' | number;
  details?: string;
};

// Normalize a variety of task object shapes into the fields used for pill mapping
export function shapeTaskForPills(t: any): ShapedTask {
  return {
    title: t?.title ?? t?.name ?? t?.summary,
    dueAt: t?.dueAt ?? t?.due_date ?? t?.due ?? t?.dueDate,
    project: t?.project ?? t?.projectName ?? t?.meta?.project ?? t?.meta?.projectName,
    link: t?.link ?? t?.url ?? t?.meta?.link,
    updatedAt: t?.updatedAt ?? t?.updated ?? t?.modifiedAt ?? t?.meta?.updated,
    priority: t?.priority ?? t?.prio ?? t?.meta?.priority,
    details: t?.details ?? t?.note ?? t?.description ?? t?.subtitle ?? t?.meta?.details,
  };
}

// In tight rows, ensure highest-signal pills appear earlier.
// Order: type, when, details, project, link, updated. Optionally truncate to `max`.
export function reorderTaskPillsForTightRows(pills: Pill[], max?: number): Pill[] {
  const desired: ReadonlyArray<Pill["kind"]> = [
    "type",
    "when",
    "details",
    "project",
    "link",
    "updated",
  ];
  const rank = (k: Pill["kind"]) => {
    const i = desired.indexOf(k);
    return i === -1 ? Number.POSITIVE_INFINITY : i;
  };
  const sorted = [...pills].sort((a, b) => rank(a.kind) - rank(b.kind));
  return typeof max === "number" ? sorted.slice(0, Math.max(1, max)) : sorted;
}

// Convenience: build and reorder pills for tight rows in one step
export function taskPillsForTightRow(
  shaped: ShapedTask,
  mapFn: (t: ShapedTask) => Pill[],
  max = 4,
) {
  return reorderTaskPillsForTightRows(mapFn(shaped), max);
}
