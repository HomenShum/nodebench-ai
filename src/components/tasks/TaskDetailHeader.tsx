import { useMemo } from "react";
import MetaPills from "../MetaPills";
import { taskToPills } from "../../lib/metaPillMappers";
import { shapeTaskForPills } from "../../lib/tasks";

export function TaskDetailHeader({ task }: { task: any }) {
  const shaped = shapeTaskForPills(task);
  const { title, dueAt, project, link, updatedAt, priority, details } = shaped;
  const pills = useMemo(
    () => taskToPills({ title, dueAt, project, link, updatedAt, priority, details }),
    [title, dueAt, project, link, updatedAt, priority, details]
  );
  return (
    <header className="space-y-2">
      <h1 className="text-lg font-semibold">{title}</h1>
      <MetaPills pills={pills} max={5} typePillClassName="border-sky-300 text-sky-700 bg-sky-100 dark:border-sky-800 dark:text-sky-200 dark:bg-sky-900/40" />
    </header>
  );
}
