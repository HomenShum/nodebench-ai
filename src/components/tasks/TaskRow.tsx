import { useMemo } from "react";
import MetaPills from "../MetaPills";
import { taskToPills } from "../../lib/metaPillMappers";
import { shapeTaskForPills, reorderTaskPillsForTightRows } from "../../lib/tasks";
import { ListTodo } from "lucide-react";

export function TaskRow({ t }: { t: any }) {
  const shaped = shapeTaskForPills(t);
  const { title, dueAt, project, link, updatedAt, priority, details } = shaped;
  const pills = useMemo(
    () => reorderTaskPillsForTightRows(taskToPills({ title, dueAt, project, link, updatedAt, priority, details })),
    [title, dueAt, project, link, updatedAt, priority, details]
  );
  return (
    <li className="task-row relative overflow-hidden flex items-start justify-between gap-3 p-2 rounded-lg hover:bg-muted/40">
      {/* Watermark: consistent row positioning (top-right) */}
      <span className="document-card__bg document-row__bg text-[var(--accent-primary)]" aria-hidden>
        <ListTodo className="h-10 w-10 rotate-12" />
      </span>
      <div className="min-w-0">
        <div className="text-sm font-medium truncate">{title}</div>
        <MetaPills pills={pills} max={4} typePillClassName="border-sky-300 text-sky-700 bg-sky-100 dark:border-sky-800 dark:text-sky-200 dark:bg-sky-900/40" />
      </div>
      <div className="shrink-0 text-xs text-muted-foreground">{/* right-side bits */}</div>
    </li>
  );
}
