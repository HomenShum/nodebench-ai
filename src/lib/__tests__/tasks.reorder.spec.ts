import { describe, it, expect } from "vitest";
import { reorderTaskPillsForTightRows } from "../tasks";
import type { Pill } from "../metaPillMappers";

const p = (kind: Pill["kind"], label: string, icon = ""): Pill => ({ kind, label, icon });

describe("reorderTaskPillsForTightRows", () => {
  it("keeps ideal order when already sorted", () => {
    const input: Pill[] = [
      p("type", "Task"),
      p("when", "Due today"),
      p("details", "P1"),
      p("project", "Alpha"),
      p("link", "example.com"),
      p("updated", "Updated 1d ago"),
    ];
    const out = reorderTaskPillsForTightRows(input);
    expect(out).toMatchSnapshot();
  });

  it("moves details before project when needed", () => {
    const input: Pill[] = [
      p("type", "Task"),
      p("when", "Due tomorrow"),
      p("project", "Beta"),
      p("details", "High"),
      p("link", "example.org"),
    ];
    const out = reorderTaskPillsForTightRows(input);
    expect(out).toMatchSnapshot();
  });

  it("handles missing type/when gracefully", () => {
    const input: Pill[] = [
      p("project", "Gamma"),
      p("details", "Low"),
      p("updated", "Updated 2h ago"),
    ];
    const out = reorderTaskPillsForTightRows(input);
    expect(out).toMatchSnapshot();
  });

  it("respects max truncation", () => {
    const input: Pill[] = [
      p("type", "Task"),
      p("project", "Delta"),
      p("details", "P2"),
      p("when", "Due next week"),
      p("link", "example.com"),
      p("updated", "Updated 5m ago"),
    ];
    const out = reorderTaskPillsForTightRows(input, 4);
    expect(out).toMatchSnapshot();
  });
});
