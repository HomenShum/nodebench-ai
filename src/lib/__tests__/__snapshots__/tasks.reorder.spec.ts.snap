// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`reorderTaskPillsForTightRows > handles missing type/when gracefully 1`] = `
[
  {
    "icon": "",
    "kind": "details",
    "label": "Low",
  },
  {
    "icon": "",
    "kind": "project",
    "label": "Gamma",
  },
  {
    "icon": "",
    "kind": "updated",
    "label": "Updated 2h ago",
  },
]
`;

exports[`reorderTaskPillsForTightRows > keeps ideal order when already sorted 1`] = `
[
  {
    "icon": "",
    "kind": "type",
    "label": "Task",
  },
  {
    "icon": "",
    "kind": "when",
    "label": "Due today",
  },
  {
    "icon": "",
    "kind": "details",
    "label": "P1",
  },
  {
    "icon": "",
    "kind": "project",
    "label": "Alpha",
  },
  {
    "icon": "",
    "kind": "link",
    "label": "example.com",
  },
  {
    "icon": "",
    "kind": "updated",
    "label": "Updated 1d ago",
  },
]
`;

exports[`reorderTaskPillsForTightRows > moves details before project when needed 1`] = `
[
  {
    "icon": "",
    "kind": "type",
    "label": "Task",
  },
  {
    "icon": "",
    "kind": "when",
    "label": "Due tomorrow",
  },
  {
    "icon": "",
    "kind": "details",
    "label": "High",
  },
  {
    "icon": "",
    "kind": "project",
    "label": "Beta",
  },
  {
    "icon": "",
    "kind": "link",
    "label": "example.org",
  },
]
`;

exports[`reorderTaskPillsForTightRows > respects max truncation 1`] = `
[
  {
    "icon": "",
    "kind": "type",
    "label": "Task",
  },
  {
    "icon": "",
    "kind": "when",
    "label": "Due next week",
  },
  {
    "icon": "",
    "kind": "details",
    "label": "P2",
  },
  {
    "icon": "",
    "kind": "project",
    "label": "Delta",
  },
]
`;
